### PR TITLE
BTRX- 268 - DUOS: Authenticate Researcher Data Access Request w/ ERA Commons ID

### PR DESCRIPTION
**Pull request by LeoForconesi**
## Addresses Jira Story:
[BTRX-268](https://broadinstitute.atlassian.net/browse/BTRX-268) : _Authenticate Researcher Data Access Request w/ ERA Commons ID_ 
## Changes in this PR:
- This story and code is related to [BTRX-268](https://github.com/DataBiosphere/consent-ui/pull/202) Ui branch
- Implement JWT authentication for shibboleth's token. We are using [JJWT](https://www.google.com.ar/search?q=jjwt&oq=jjwt&aqs=chrome..69i57j0j69i60l3j69i59.2231j0j9&sourceid=chrome&ie=UTF-8) library for this purpose
- Save, Map and return ERA Commons values, such as Authorized status, Expiration Date and Nih User Name
- Modified logic of how a Researcher Profile is updated. This way we can update without deleting all properties by User Id
- Replace Data Access Requests values that were mapped from MySql to MongoDb so later in the Ui we show a snapshot of these values at the moment of its creation
- Some minor code refactor.

_UPDATE_
- We have included PDFs missing data fields corresponding to [BTRX-289](https://broadinstitute.atlassian.net/browse/BTRX-289) - Missing information in downloadable Full DAR Application PDF 

_NOTE_
- Before merging we need to add Nih Jwt secret to Vault (test & prod) and modify Docker config template version to include the new yaml field.

---

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: consent-ws/src/main/java/org/broadinstitute/consent/http/db/ResearcherPropertyDAO.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/db/ResearcherPropertyDAO.java]: Nit: typo, `Propertie` should be pluralized `deletePropertiesByUserAndKey`

File: consent-ws/src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java]: Nit: You probably don't need this regex in the path. We used this technique for handling basic authentication and I think the pattern stuck. It can probably just be `@Path("api/nih-login/")`. This should be true in lots of other cases too.

File: consent-ws/src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java]: Nit: spacing here and on the catch line as well.

File: consent-ws/src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java]: Nit: several extra lines here, and below.

File: consent-ws/src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java]: Nit, naming. Should this be `updateEraExpirationDates` instead? Also, the `update` doesn't make sense here because we're not actually updating, just generating a new map with a single date value.

File: consent-ws/src/main/java/org/broadinstitute/consent/http/service/users/handler/DatabaseResearcherAPI.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/service/users/handler/DatabaseResearcherAPI.java]: Nice spacing fix, but need one before the `{` as well.

File: consent-ws/src/main/java/org/broadinstitute/consent/http/service/users/handler/DatabaseResearcherAPI.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/service/users/handler/DatabaseResearcherAPI.java]: Nit: extra space.

**Comment by rushtong** [comment]
Do you want to document this in `api-docs.yaml` or is that intentionally not there?

**Comment by LeoForconesi** [comment]
Thanks for your review Greg! I just updated `api-docs.yaml`, i was missing that.

**Comment by rushtong** [comment]
Caveat, I haven't tested anything functionally, so, pending others' approval, 👍

**Comment by LeoForconesi** [comment]
@ahaessly 
We have updated this PR. When you have some moment, please review.

**Comment by LeoForconesi** [comment]
@rushtong if you want to take a look at this code again it would really helpfull. It hasn't changed a lot though since the last time you reviewed it, only develop branch updates and adding pdf corresponding fields.

**Comment by rushtong** [comment]
With respect to `nihAuthentication`, we'll need a corresponding change to the config similar to the one we did for the [dataset configs](https://github.com/broadinstitute/firecloud-develop/pull/1408)

As a practical note, is there a way we can write the `NihConfiguration` such that it can be null in the configs in the deployed app? If we can do that, and log (log level severe) the fact that it doesn't exist, then the consent PR and the fc-dev PR can be decoupled. This would be a nice approach for any config change. Currently, DropWizard config reader complains if there's any kind of mis-match at all, likely because it's using jackson to model the config class alongside the config file. It would be a *lot* easier to deploy if this can be done.

File: consent-ws/src/main/java/org/broadinstitute/consent/http/resources/DataRequestReportsResource.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/resources/DataRequestReportsResource.java]: Could be a lot more compact with a stream: 
```
        return Arrays.stream(fieldsForManualReview).
            filter(form::containsKey).
            map(field -> Boolean.valueOf(form.get(field).toString())).
            collect(Collectors.toSet()).
            contains(Boolean.TRUE);
```

File: consent-ws/src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java]: unused

**Comment by rushtong** [comment]
FYI, I have not looked into this, but we should do a security review due to JWT. Please review https://www.owasp.org/index.php/JSON_Web_Token_(JWT)_Cheat_Sheet_for_Java for any potential issues. I'll ping our devops security to see if they want to review.

**Comment by rushtong** [comment]
@adrazhi - Can I ask for your review on this due to JWT? Let's meet to discuss if you think we need to.

File: consent-ws/src/main/java/org/broadinstitute/consent/http/resources/Resource.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/resources/Resource.java]: I question whether all of these are really bad requests. Some seem like they should at least be internal server errors. Additionally, I think it might make more sense to catch these in `NihServiceAPI` instead and log them since we should be alerted via sentry if we see these kinds of errors.

File: consent-ws/src/main/java/org/broadinstitute/consent/http/service/DataAccessParser.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/service/DataAccessParser.java]: Any reason we're not using constants here and below?

File: consent-ws/src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java
Comment (rushtong) [inline consent-ws/src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java]: See my comment about exceptions above. I think this block should catch and log (probably at the warning level for client-provided `jwt` tokens) all of the exceptions possible with some explanatory language for each.

File: consent-ws/src/main/resources/assets/api-docs.yaml
Comment (rushtong) [inline consent-ws/src/main/resources/assets/api-docs.yaml]: nit: Technically, it's `eRA Commons`. See https://public.era.nih.gov/commons/public/login.do and  https://era.nih.gov/

**Comment by rushtong** [comment]
With respect to `RequestApplication.pdf` - instead of checking that into code, I think a better approach is to put it in a consent-accessible bucket and render it via the GCS api. This way we don't need to do a code release for a simple change in the template file.

File: consent-ws/src/test/java/org/broadinstitute/consent/http/NihAccountTest.java
Comment (rushtong) [inline consent-ws/src/test/java/org/broadinstitute/consent/http/NihAccountTest.java]: nit: unnecessary extra lines

File: consent-ws/src/test/java/org/broadinstitute/consent/http/NihAccountTest.java
Comment (rushtong) [inline consent-ws/src/test/java/org/broadinstitute/consent/http/NihAccountTest.java]: I don't have a suggestion, but would love to see if there's a way to test a security scenario. Will try to spend some time digesting the OWASP doc and see if there's something more we can look for.

**Review by rushtong — CHANGES_REQUESTED** [review]
Looks good, but I definitely have some concerns that are beyond my knowledge at the moment.

File: consent-ws/pom.xml
Comment (rushtong) [inline consent-ws/pom.xml]: Latest version here is `0.10.5` ... there have been a bunch of releases in August.

File: consent-ws/src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java
Comment (rushtong) [reply consent-ws/src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java]: See the [quickstart example](https://github.com/jwtk/jjwt#quickstart): 
```
try {
    Jwts.parser().setSigningKey(key).parseClaimsJws(compactJws);
    //OK, we can trust this JWT
} catch (JwtException e) {
    //don't trust the JWT!
}
```
In our implementation, `parsePlaintextJws` is where we care about catching any kind of error. This is where we'll find any kind of security problem with the token. These should be logged as severe (with some more info about user, request, etc.,) so we can follow up with those users via our normal communications process.

**Comment by adrazhi** [comment]
> @adrazhi - Can I ask for your review on this due to JWT? Let's meet to discuss if you think we need to.

@rushtong 
So the first concern related to JWT is the integrity protection and to ensure JWT is protected by either a signature or a MAC. Do not allow the unsecured JWTs: {"alg":"none"}.  I think you're going the signature direction which is great and I think it's preferred over HMAC. 

Few other things I'd recommend watching out for: 
1. Use a cryptographically random JWT secret. You are using [this library](https://github.com/jwtk/jjwt) to do that, correct?
2. What is the current token expiration time? 
3. Obvious but worth mentioning to not store any sensitive data in the JWT payload as it can be decoded [easily.](https://jwt.io/#debugger-io) 
4. On headers: Ensure embedded "jwk elements are not trusted". 

Let me know if you have any questions.

File: src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java]: nit: extra line

File: src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/resources/NihAccountResource.java]: I'm still worried that this takes a userid as a path param. I think it would be much safer to extract this using `@Auth User user`

File: src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/NihServiceAPI.java]: nit: extra line

File: src/main/resources/assets/api-docs.yaml
Comment (rushtong) [inline src/main/resources/assets/api-docs.yaml]: nit: I can't remember this or find it in the comments, but I thought we discussed making this just `/api/nih` instead of `/api/nih-register`

**Review by rushtong — CHANGES_REQUESTED** [review]
Two change requests, some nits, otherwise looks good!

**Comment by LeoForconesi** [comment]
Closing this PR. Review comments are solved in [PR-309](https://github.com/DataBiosphere/consent/pull/309).

**Pull request closed without merging**
